types/d3-array/d3-array-tests.ts

```diff
@@ -380,11 +567,17 @@ numbersArray = d3Array.range(1, 10, 0.5);
 // shuffle() -------------------------------------------------------------------
 
 mergedArray = d3Array.shuffle(mergedArray);
-
 mergedArray = d3Array.shuffle(mergedArray, 1);
-
 mergedArray = d3Array.shuffle(mergedArray, 1, 3);
 
+mergedArray = d3Array.shuffle(readonlyMergedArray);
```

Comment:
You certainly have a point regarding the unusual number of overloads.

However, one issue is that one would expect that the return type matches the argument type passed in. I would be hesitant to actually pass in `T[]` and be returned `ArrayLike<T>`, similar with Typed Array inputs.

So I think, it might be better to simply bite the overload bullet here. An alternative with only one, yet cumbersome, overload would be something like:

```ts
export function shuffle<TA extends Float32Array | Float64Array | ...>(array: TA, lo?: number, hi?: number): TA;
```
The downside with this approach is, that we know changed the semantics of the generic from element type to array type for this overloaded signature.